packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts

```diff
@@ -4,33 +4,72 @@
  */
 
 import { ITelemetryLogger } from "@fluidframework/common-definitions";
-import { assert } from "@fluidframework/common-utils";
+import { assert, LazyPromise } from "@fluidframework/common-utils";
 import { cloneGCData } from "@fluidframework/garbage-collector";
+import { ISnapshotTree, SummaryType } from "@fluidframework/protocol-definitions";
 import {
+    CreateChildSummarizerNodeParam,
+    gcBlobKey,
     IContextSummarizeResult,
+    IGCData,
+    IGCDetails,
     ISummarizeInternalResult,
     ISummarizerNodeConfig,
     ISummarizerNodeWithGC,
-    CreateChildSummarizerNodeParam,
-    IGCData,
 } from "@fluidframework/runtime-definitions";
 import { SummarizerNode } from "./summarizerNode";
-import { ICreateChildDetails, IInitialSummary, ISummarizerNodeRootContract, SummaryNode } from "./summarizerNodeUtils";
+import {
+    EscapedPath,
+    ICreateChildDetails,
+    IInitialSummary,
+    ISummarizerNodeRootContract,
+    ISummaryNode,
+    ReadAndParseBlob,
+    SummaryNode,
+} from "./summarizerNodeUtils";
 
 export interface IRootSummarizerNodeWithGC extends ISummarizerNodeWithGC, ISummarizerNodeRootContract {}
 
+interface ISummaryNodeWithGC extends ISummaryNode {
+    readonly used: boolean;
+}
+
+// Extend SummaryNode to add used state tracking to it.
+class SummaryNodeWithGC extends SummaryNode implements ISummaryNodeWithGC {
+    constructor(
+        public readonly used: boolean,
+        summary: {
+            readonly referenceSequenceNumber: number,
+            readonly basePath: EscapedPath | undefined,
+            readonly localPath: EscapedPath,
+            additionalPath?: EscapedPath,
+        },
+    ) {
+        super(summary);
+    }
+}
+
 /**
  * Extends the functionality of SummarizerNode to manage this node's garbage collection data:
  * - Adds a new API `getGCData` to return GC data of this node.
  * - Caches the result of getGCData method to be used if nothing changes between summaries.
- * - Gets the initial GC data if required.
  * - Adds GC data to the result of summarize.
  * - Adds trackState param to summarize. If trackState is false, it bypasses the SummarizerNode and calls
  *   directly into summarizeInternal method.
  */
 export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummarizerNodeWithGC {
+    public used: boolean = false;
     private gcData: IGCData | undefined;
 
+    // Tracks the work-in-progress used state during summary.
+    private wipUsedState: boolean | undefined;
```

Comment:
I'd prefer enum (similar to detached / attaching / attached). That makes code easier to follow, and compiler more likely to tell us when we have bugs in code.

Reply:
same below

Reply:
I don't prefer enum here, but if you want to be more explicit about undefined meaning you could use a literal union type, something like `"referenced" | "unreferenced" | "unknown"`, but then I would like to map to the actual bool types needed and so it becomes: `boolean | "unknown"`, but in this case I feel like `undefined` actually has the same/better meaning, yielding the same end result: `boolean | undefined`.

Main reason I feel like this is different from detached/attaching/attached state object, because this is "work in progress" state that only exists while summarizing.

I do feel like encapsulating all wip state in an object would be better though. I just think that the wip state object for the extended class would end up being a separate object (for simplicity) from the base class wip object, which then means it would only be this one flag for now.

Reply:
I think there was some confusion as to what this is because I named it "state". Its not really a state, its just a value. I am using used routes here and not a boolean so that would clear things up.